packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1444,7 +1446,11 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
     // Perform work on root as if the given expiration time is the current time.
     // This has the effect of synchronously flushing all work up to and
     // including the given time.
-    performWorkOnRoot(root, expirationTime, false);
+    do {
+      performWorkOnRoot(root, expirationTime, false);
+      findHighestPriorityRoot();
+    } while (nextFlushedRoot === root && nextFlushedExpirationTime === Sync);
```

Comment:
Loops are scary. Are we sure we need this?

Reply:
Typically `performWorkOnRoot` is called inside a loop that finds next thing to work on until there's none. `flushRoot` is special because it wants to flush "just one" root. Which is also why we forgot to flush `setState` in commit lifecycles (since they're just like flushing the same root twice—if my mental model is right).

There could be more than one cascading update. So it seems like a loop is necessary.

I'm not sure those are the right exit conditions though.

Reply:
Overall I'm not feeling confident in this. There are other things `performWork` does that `performWorkOnRoot` doesn’t. For example scheduling a deferred callback for leftover work. I’m not sure if missing this here means it would get dropped. I wonder if we could reuse the rest of the scheduler code, but teach the scheduler the concept of the "current" batch which gets committed in isolation if it exists.

Reply:
Also not sure what would happen if we used `flushSync` to schedule another root inside a currently committed batch's lifecycle.